### ISSUE TITLE
utils.help_search looks for /doc/KEYWORDS.db in the wrong directory.

### ISSUE THREAD

**Issue by joh4nd** [rpy2/rpy2#1033]
Following the answer to fix a bug https://github.com/rpy2/rpy2/issues/1024 I `pip install rpy2` to the newest version.

python returned:
- `R[write to console]: Error in file(con, "r") : cannot open the connection`
- `R[write to console]:  cannot open file '/usr/lib/R/doc/KEYWORDS.db': No such file or directory`

The hsearch-utils package in R refers to [hsearch_db](https://rdrr.io/r/utils/hsearch-utils.html). rpy2 seems to look for KEYWORDS.db in a wrong place. There is no `doc` folder in `/usr/lib/R/`. Fortunately I found out that `KEYWORDS.db` **actually is in: `/usr/share/R/doc/`**.

**inspect help_search**

```python
print(rutils.help_search)

function (pattern, fields = c("alias", "concept", "title"), apropos, 
    keyword, whatis, ignore.case = TRUE, package = NULL, lib.loc = NULL, 
    help.db = getOption("help.db"), verbose = getOption("verbose"), 
    rebuild = FALSE, agrep = NULL, use_UTF8 = FALSE, types = getOption("help.search.types")) 
{
    .wrong_args <- function(args) gettextf("argument %s must be a single character string", 
        sQuote(args))
    if (is.logical(verbose)) 
        verbose <- 2 * as.integer(verbose)
    fuzzy <- agrep
    if (!missing(pattern)) {
        if (!is.character(pattern) || (length(pattern) > 1L)) 
            stop(.wrong_args("pattern"), domain = NA)
        i <- pmatch(fields, hsearch_db_fields)
        if (anyNA(i)) 
            stop("incorrect field specification")
        else fields <- hsearch_db_fields[i]
    }
    else if (!missing(apropos)) {
        if (!is.character(apropos) || (length(apropos) > 1L)) 
            stop(.wrong_args("apropos"), domain = NA)
        else {
            pattern <- apropos
            fields <- c("alias", "title")
        }
    }
    else if (!missing(keyword)) {
        if (!is.character(keyword) || (length(keyword) > 1L)) 
            stop(.wrong_args("keyword"), domain = NA)
        else {
            pattern <- keyword
            fields <- "keyword"
            if (is.null(fuzzy)) 
                fuzzy <- FALSE
        }
    }
    else if (!missing(whatis)) {
        if (!is.character(whatis) || (length(whatis) > 1)) 
            stop(.wrong_args("whatis"), domain = NA)
        else {
            pattern <- whatis
            fields <- "alias"
        }
    }
    else {
        stop("do not know what to search")
    }
    if (!missing(help.db)) 
        warning("argument 'help.db' is deprecated")
    i <- pmatch(types, hsearch_db_types)
    if (anyNA(i)) 
        stop("incorrect type specification")
    else types <- hsearch_db_types[i]
    db <- hsearch_db(package, lib.loc, types, verbose, rebuild, 
        use_UTF8)
    lib.loc <- attr(db, "LibPaths")
    if (!identical(sort(types), sort(attr(db, "Types")))) {
        db$Base <- db$Base[!is.na(match(db$Base$Type, types)), 
            ]
        db[-1L] <- lapply(db[-1L], function(e) {
            e[!is.na(match(e$ID, db$Base$ID)), ]
        })
    }
    if (!is.null(package)) {
        pos_in_hsearch_db <- match(package, unique(db$Base[, 
            "Package"]), nomatch = 0L)
        if (any(pos_in_hsearch_db) == 0L) 
            stop(gettextf("no information in the database for package %s: need 'rebuild = TRUE'?", 
                sQuote(package[pos_in_hsearch_db == 0][1L])), 
                domain = NA)
        db[] <- lapply(db, function(e) {
            e[!is.na(match(e$Package, package)), ]
        })
    }
    if (verbose >= 2L) {
        message("Database of ", NROW(db$Base), " help objects (", 
            NROW(db$Aliases), " aliases, ", NROW(db$Concepts), 
            " concepts, ", NROW(db$Keywords), " keywords)", domain = NA)
        flush.console()
    }
    if (!length(db$Base)) 
        return(invisible())
    if (is.null(fuzzy) || is.na(fuzzy)) 
        fuzzy <- (grepl("^([[:alnum:]]|[[:space:]]|-)+$", pattern) && 
            (nchar(pattern, type = "c") > 4L))
    if (is.logical(fuzzy)) {
        if (fuzzy) 
            max.distance <- 0.1
    }
    else if (is.numeric(fuzzy) || is.list(fuzzy)) {
        max.distance <- fuzzy
        fuzzy <- TRUE
    }
    else stop("incorrect 'agrep' specification")
    dbBase <- db$Base
    search_fun <- if (fuzzy) {
        function(x) {
            agrep(pattern, x, ignore.case = ignore.case, max.distance = max.distance)
        }
    }
    else {
        function(x) {
            grep(pattern, x, ignore.case = ignore.case, perl = use_UTF8)
        }
    }
    search_db_results <- function(p, f, e) data.frame(Position = p, 
        Field = f, Entry = e, stringsAsFactors = FALSE)
    search_db_field <- function(field) {
        switch(field, alias = {
            aliases <- db$Aliases$Alias
            matched <- search_fun(aliases)
            search_db_results(match(db$Aliases$ID[matched], dbBase$ID), 
                rep.int(field, length(matched)), aliases[matched])
        }, concept = {
            concepts <- db$Concepts$Concept
            matched <- search_fun(concepts)
            search_db_results(match(db$Concepts$ID[matched], 
                dbBase$ID), rep.int(field, length(matched)), 
                concepts[matched])
        }, keyword = {
            keywords <- db$Keywords$Keyword
            matched <- search_fun(keywords)
            search_db_results(match(db$Keywords$ID[matched], 
                dbBase$ID), rep.int(field, length(matched)), 
                keywords[matched])
        }, name = {
            matched <- search_fun(dbBase$Name)
            search_db_results(matched, rep.int("Name", length(matched)), 
                dbBase$Name[matched])
        }, title = {
            matched <- search_fun(dbBase$Title)
            search_db_results(matched, rep.int("Title", length(matched)), 
                dbBase$Title[matched])
        })
    }
    matches <- NULL
    for (f in fields) matches <- rbind(matches, search_db_field(f))
    matches <- matches[order(matches$Position), ]
    db <- cbind(dbBase[matches$Position, c("Topic", "Title", 
        "Name", "ID", "Package", "LibPath", "Type"), drop = FALSE], 
        matches[c("Field", "Entry")])
    rownames(db) <- NULL
    if (verbose >= 2L) {
        n_of_objects_matched <- length(unique(db[, "ID"]))
        message(sprintf(ngettext(n_of_objects_matched, "matched %d object.", 
            "matched %d objects."), n_of_objects_matched), domain = NA)
        flush.console()
    }
    y <- list(pattern = pattern, fields = fields, type = if (fuzzy) "fuzzy" else "regexp", 
        agrep = agrep, ignore.case = ignore.case, types = types, 
        package = package, lib.loc = lib.loc, matches = db)
    class(y) <- "hsearch"
    y
}
<bytecode: 0x2d3e1e8>
<environment: namespace:utils>
```

**To reproduce**

```python
from rpy2.robjects.packages import importr
from rpy2.robjects import r
from rpy2.robjects import pandas2ri 
rutils = importr('utils')
rhelp_where = rutils.help_search('help')
```

**Error**

```python
# console/stdout

R[write to console]: Error in file(con, "r") : cannot open the connection

R[write to console]: In addition: 
R[write to console]: Warning messages:

R[write to console]: 1: 
R[write to console]: In (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :
R[write to console]: 
 
R[write to console]:  libraries '/usr/local/lib/R/site-library', '/usr/lib/R/site-library' contain no packages

R[write to console]: 2: 
R[write to console]: In file(con, "r") :
R[write to console]: 
 
R[write to console]:  cannot open file '/usr/lib/R/doc/KEYWORDS.db': No such file or directory

---------------------------------------------------------------------------
RRuntimeError               Traceback (most recent call last)
/home/this/path/to/src/multiple_imputation.py in line 1
----> 8 rhelp_where = rutils.help_search('help')

File ~/path/to/bin/venv/lib/python3.9/site-packages/rpy2/robjects/functions.py:208, in SignatureTranslatedFunction.__call__(self, *args, **kwargs)
    206         v = kwargs.pop(k)
    207         kwargs[r_k] = v
--> 208 return (super(SignatureTranslatedFunction, self)
    209         .__call__(*args, **kwargs))

File ~/path/to/bin/venv/lib/python3.9/site-packages/rpy2/robjects/functions.py:131, in Function.__call__(self, *args, **kwargs)
    129     else:
    130         new_kwargs[k] = cv.py2rpy(v)
--> 131 res = super(Function, self).__call__(*new_args, **new_kwargs)
    132 res = cv.rpy2py(res)
    133 return res

File ~/path/to/bin/venv/lib/python3.9/site-packages/rpy2/rinterface_lib/conversion.py:45, in _cdata_res_to_rinterface.._(*args, **kwargs)
     44 def _(*args, **kwargs):
---> 45     cdata = function(*args, **kwargs)
     46]     # TODO: test cdata is of the expected CType
     47     return _cdata_to_rinterface(cdata)

File ~/path/to/bin/venv/lib/python3.9/site-packages/rpy2/rinterface.py:860, in SexpClosure.__call__(self, *args, **kwargs)
    853     res = rmemory.protect(
    854         openrlib.rlib.R_tryEval(
    855             call_r,
    856             call_context.__sexp__._cdata,
    857             error_occured)
    858    )
    859    if error_occured[0]:
--> 860        raise embedded.RRuntimeError(_rinterface._geterrmessage())
    861 return res

RRuntimeError: Error in file(con, "r") : cannot open the connection
```

**Comment by eddelbuettel** [comment]
This also made it to the r-sig-debian list.  First message of the thread it spawned as indexed [by mail-archive.com](https://www.mail-archive.com/r-sig-debian@r-project.org/msg03353.html) as well as via the [official mailman archive](https://stat.ethz.ch/pipermail/r-sig-debian/2023-June/003464.html).

**Comment by lgautier** [comment]
This is odd. `rpy2` does not specify that path.

I see. R is specifying this through the environment variable `R_DOC_DIR` in the R shell script that is a front-end to the R binary. That script is creating a build/install time. Other software embedding R will have the same issue as `rpy2`.

An additional peculiarity that of those variables are in the binary executable rather than left to the script. In list below I think that all but `R_HOME` are defined in the R binary.

```python

In [4]: print(rpy2.robjects.r('local({envs <- Sys.getenv(); envs[grepl("^R_", names(envs))]
   ...: })'))
R[write to console]: Error in formatDL(x, ...) : unused argument (useS4 = FALSE)

/home/laurent/Desktop/software/python/rpy2_github/rpy2/robjects/robject.py:123: UserWarning: Invalid call to "show()" in R: Error in formatDL(x, ...) : unused argument (useS4 = FALSE)

  warnings.warn(f'Invalid call to "show()" in R: {rre}')
R_BROWSER               /usr/bin/firefox
R_BZIPCMD               /usr/bin/bzip2
R_GZIPCMD               /usr/bin/gzip
R_HOME                  /usr/local/packages/R/4.2/lib/R
R_LIBS_SITE             /usr/local/packages/R/4.2/lib/R/site-library
R_LIBS_USER             /home/laurent/R/x86_64-pc-linux-gnu-library/4.2
R_PAPERSIZE             letter
R_PDFVIEWER             /usr/bin/xdg-open
R_PLATFORM              x86_64-pc-linux-gnu
R_PRINTCMD              lpr
R_RD4PDF                times,inconsolata,hyper
R_SESSION_TMPDIR        /tmp/RtmpZol5si
R_STRIP_SHARED_LIB      strip --strip-unneeded
R_STRIP_STATIC_LIB      strip --strip-debug
R_SYSTEM_ABI            linux,gcc,gxx,gfortran,gfortran
R_TEXI2DVICMD           texi2dvi
R_UNZIPCMD              /usr/bin/unzip
R_ZIPCMD                /usr/bin/zip

```

**Comment by lgautier** [comment]
In the conversation on the debian list shared by @eddelbuettel it is mentioned that the link to the R documentation should be obtained with:

```r
> file.path( R.home("doc"), "KEYWORDS.db" )
   [1] "/usr/share/R/doc/KEYWORDS.db"
 ```

```r
> R.home
function (component = "home") 
{
    rh <- .Internal(R.home())
    switch(component, home = rh, bin = if (.Platform$OS.type == 
        "windows" && nzchar(p <- .Platform$r_arch)) file.path(rh, 
        component, p) else file.path(rh, component), share = if (nzchar(p <- Sys.getenv("R_SHARE_DIR"))) p else file.path(rh, 
        component), doc = if (nzchar(p <- Sys.getenv("R_DOC_DIR"))) p else file.path(rh, 
        component), include = if (nzchar(p <- Sys.getenv("R_INCLUDE_DIR"))) p else file.path(rh, 
        component), modules = if (nzchar(p <- .Platform$r_arch)) file.path(rh, 
        component, p) else file.path(rh, component), file.path(rh, 
        component))
}
```

It is indeed defined by `R_DOC_DIR` that is only defined in the `R` script:

**Comment by lgautier** [comment]
I see that the comment above was already made in the discussion on the debian list.

**Comment by lgautier** [comment]
Ok, this boils down to finding a way to get `R_DOC_DIR` from the R binary. 

`R CMD config` does not seem able to communicate what are the variables in defined in the R frontend script. This leaves a) parsing the frontend script R, or b) running  `R -s -e 'envs <- Sys.getenv(); envs[grepl("^R_", names(envs))]'` and parse the output.

**Comment by eddelbuettel** [comment]
Two quick comments:  

First, on both Ubuntu and Debian a quick `Rscript -e 'Sys.getenv("R_DOC_DIR")'` gets me the correct value so it is odd that @joh4nd didn't get the correct value.

Second, for RInside and littler I cheat and detect the environment variables at _build_ time and make them compile-time constants.  Oddly enough that somewhat brute-force approach survived all those years.

**Comment by lgautier** [comment]
Fixed with PR #1037.
